File: trunk/managed/Warp3D/Warp3DVB/warp_Color.vb
' Created by: X
' http://www.createdbyx.com/
' ----------------------
' C# to VB.NET conversion of this code was done by hand and with the assistance of the 
' Convert C# to VB .NET application availible at http://www.kamalpatel.net/ConvertCSharp2VB.aspx
' ConvertCSharp2VB utility was developed by: Kamal Patel (http://www.KamalPatel.net) 
' ----------------------
' Thanks to Alan Simes ( http://alansimes.blogdns.net/ ) for creating the Warp3D library.
' http://alansimes.blogdns.net/cs/files/default.aspx
' ----------------------

Imports System

Namespace Rednettle.Warp3D
    '/ <summary>
    '/ Summary description for warp_Color.
    '/ </summary>
    Public Class warp_Color
        Public Shared ALPHA As Integer = &HFF000000  ' alpha mask
        Public Shared RED As Integer = &HFF0000  ' red mask
        Public Shared GREEN As Integer = &HFF00  ' green mask
        Public Shared BLUE As Integer = &HFF  ' blue mask
        Public Shared MASK7Bit As Integer = &HFEFEFF  ' mask for additive/subtractive shading
        Public Shared MASK6Bit As Integer = &HFCFCFC  ' mask for additive/subtractive shading
        Public Shared RGB As Integer = &HFFFFFF  ' rgb mask

        Private Shared pixel As Integer, color As Integer, overflow As Integer, colorscale As Integer, r As Integer, g As Integer, b As Integer

        Private Sub New()
        End Sub

        Public Shared Function random(ByVal color As Integer, ByVal delta As Integer) As Integer
            Dim rnd As Random = New Random()

            Dim r As Integer = (color >> 16) And 255
            Dim g As Integer = (color >> 8) And 255
            Dim b As Integer = color And 255

            r = CInt(r + (rnd.NextDouble() * delta))
            g = CInt(g + (rnd.NextDouble() * delta))
            b = CInt(b + (rnd.NextDouble() * delta))

            Return getCropColor(r, g, b)
        End Function

        Public Shared Function random() As Integer
            Dim rnd As Random = New Random()
            Return CInt((rnd.NextDouble() * 16777216))
        End Function

        Public Shared Function getRed(ByVal c As Integer) As Integer
            Return (c And RED) >> 16
        End Function

        Public Shared Function getGreen(ByVal c As Integer) As Integer
            Return (c And GREEN) >> 8
        End Function

        Public Shared Function getBlue(ByVal c As Integer) As Integer
            Return c And BLUE
        End Function

        Public Shared Function makeGradient(ByVal colors() As Integer, ByVal size As Integer) As Integer()
            Dim pal(size - 1) As Integer ' = New Integer(size) {}
            Dim c1 As Integer, c2 As Integer, pos1 As Integer, pos2 As Integer, range As Integer
            Dim r As Integer, g As Integer, b As Integer, r1 As Integer, g1 As Integer, b1 As Integer, r2 As Integer, g2 As Integer, b2 As Integer, dr As Integer, dg As Integer, db As Integer
            If colors.GetLength(0) = 1 Then
                c1 = colors(0)

                For i As Integer = 0 To size - 1 'Step i + 1
                    pal(i) = c1
                Next
                Return pal
            End If


            For c As Integer = 0 To colors.GetLength(0) - 2 ' Step c + 1
                c1 = colors(c)
                c2 = colors(c + 1)
                pos1 = size * c \ (colors.GetLength(0) - 1)
                pos2 = size * (c + 1) \ (colors.GetLength(0) - 1)
                range = pos2 - pos1
                r1 = warp_Color.getRed(c1) << 16
                g1 = warp_Color.getGreen(c1) << 16
                b1 = warp_Color.getBlue(c1) << 16
                r2 = warp_Color.getRed(c2) << 16
                g2 = warp_Color.getGreen(c2) << 16
                b2 = warp_Color.getBlue(c2) << 16
                dr = (r2 - r1) \ range
                dg = (g2 - g1) \ range
                db = (b2 - b1) \ range
                r = r1
                g = g1
                b = b1

                For i As Integer = pos1 To pos2 - 1 'Step i + 1
                    pal(i) = getColor(r >> 16, g >> 16, b >> 16)
                    r += dr
                    g += dg
                    b += db
                Next
            Next

            Return pal
        End Function

        Public Shared Function add(ByVal color1 As Integer, ByVal color2 As Integer) As Integer
            pixel = (color1 And MASK7Bit) + (color2 And MASK7Bit)
            overflow = pixel And &H1010100
            overflow = overflow - (overflow >> 8)

            Return ALPHA Or overflow Or pixel
        End Function

        ' TOTO: Make sure using xor works properly xor replaces C# ~ operator
        Public Shared Function [sub](ByVal color1 As Integer, ByVal color2 As Integer) As Integer
            pixel = (color1 And MASK7Bit) + ((color2 Xor color2) And MASK7Bit)
            overflow = (pixel Xor pixel) And &H1010100
            overflow = overflow - (overflow >> 8)
            Return ALPHA Or ((overflow Xor overflow) And pixel)
        End Function

        Public Shared Function subneg(ByVal color1 As Integer, ByVal color2 As Integer) As Integer
            pixel = (color1 And MASK7Bit) + (color2 And MASK7Bit)
            overflow = (pixel Xor pixel) And &H1010100
            overflow = overflow - (overflow >> 8)
            Return ALPHA Or ((overflow Xor overflow) And pixel)
        End Function

        Public Shared Function inv(ByVal color As Integer) As Integer
            Return ALPHA Or (color Xor color)
        End Function

        Public Shared Function mix(ByVal color1 As Integer, ByVal color2 As Integer) As Integer
            ' Returns the averidge color from 2 colors
            Return ALPHA Or (((color1 And MASK7Bit) >> 1) + ((color2 And MASK7Bit) >> 1))
        End Function

        Public Shared Function scale(ByVal color As Integer, ByVal factor As Integer) As Integer
            If factor = 0 Then Return 0
            If factor = 255 Then Return color
            If factor = 127 Then Return (color And &HFEFEFE) >> 1

            r = (((color >> 16) And 255) * factor) >> 8
            g = (((color >> 8) And 255) * factor) >> 8
            b = ((color And 255) * factor) >> 8
            Return ALPHA Or (r << 16) Or (g << 8) Or b
        End Function

        Public Shared Function multiply(ByVal color1 As Integer, ByVal color2 As Integer) As Integer
            If (color1 And RGB) = 0 Then Return 0
            If (color2 And RGB) = 0 Then Return 0
            r = (((color1 >> 16) And 255) * ((color2 >> 16) And 255)) >> 8
            g = (((color1 >> 8) And 255) * ((color2 >> 8) And 255)) >> 8
            b = ((color1 And 255) * (color2 And 255)) >> 8
            Return ALPHA Or (r << 16) Or (g << 8) Or b
        End Function

        Public Shared Function transparency(ByVal bkgrd As Integer, ByVal color As Integer, ByVal alpha As Integer) As Integer
            If alpha = 0 Then Return color
             If alpha = 255 Then Return bkgrd
            If alpha = 127 Then Return mix(bkgrd, color)

            r = (alpha * (((bkgrd >> 16) And 255) - ((color >> 16) And 255)) >> 8) + ((color >> 16) And 255)
            g = (alpha * (((bkgrd >> 8) And 255) - ((color >> 8) And 255)) >> 8) + ((color >> 8) And 255)
            b = (alpha * ((bkgrd And 255) - (color And 255)) >> 8) + (color And 255)

            Return alpha Or (r << 16) Or (g << 8) Or b
        End Function

        Public Shared Function getCropColor(ByVal r As Integer, ByVal g As Integer, ByVal b As Integer) As Integer
            Return ALPHA Or (warp_Math.crop(r, 0, 255) << 16) Or (warp_Math.crop(g, 0, 255) << 8) Or warp_Math.crop(b, 0, 255)
        End Function


        Public Shared Function getColor(ByVal r As Integer, ByVal g As Integer, ByVal b As Integer) As Integer
            Return ALPHA Or (r << 16) Or (g << 8) Or b
        End Function

        Public Shared Function getGray(ByVal color As Integer) As Integer
            Dim r As Integer = ((color And RED) >> 16)
            Dim g As Integer = ((color And GREEN) >> 8)
            Dim b As Integer = (color And BLUE)
            Dim Y As Integer = CInt((r * 3 + g * 6 + b) / 10)
            Return ALPHA Or (Y << 16) Or (Y << 8) Or Y
        End Function

        Public Shared Function getAverage(ByVal color As Integer) As Integer
            Return (((color And RED) >> 16) + ((color And GREEN) >> 8) + (color And BLUE)) \ 3
        End Function
    End Class
End Namespace

File: trunk/managed/Warp3Dmod/warp_Color.cs
using System;
using System.Runtime.InteropServices;

namespace Warp3D
{
    /// <summary>
    /// Summary description for warp_Color.
    /// </summary>
    public class warp_Color
    {
        public static int MASKALPHA = unchecked((int)0xFF000000); // alpha mask
        public static int MASKHALFALPHA = unchecked((int)0x7F000000); // alpha mask
        public static int MASKRED = unchecked((int)0xFF0000);  // red mask
        public static int MASKGREEN = unchecked((int)0xFF00);  // green mask
        public static int MASKBLUE = unchecked((int)0xFF);  // blue mask
        public static int MASKRB = unchecked((int)0x00FF00FF);  // mask for additive/subtractive shading
        public static int MASKAG = unchecked((int)0xFF00FF00);  // mask for additive/subtractive shading
        public static int RBHALF = unchecked((int)0x00800080);
        public static int MASKRBHALF = unchecked((int)0xFF80FF80);
        public static int MASKRBOVER = unchecked((int)0x01000100);  
        public static int MASK7Bit = unchecked((int)0x00FEFEFF);  // mask for additive/subtractive shading
        public static int MASK6Bit = unchecked((int)0x00FCFCFC);  // mask for additive/subtractive shading
        public static int MASK7BitA = unchecked((int)0x00FEFEFF);  // mask for additive/subtractive shading
        public static int MASK6BitA = unchecked((int)0x00FCFCFC);  // mask for additive/subtractive shading
        public static int MASKRGB = unchecked((int)0x00FFFFFF);  // rgb mask

        public static int White = unchecked((int)0xFFFFFFFF);
        public static int Grey = unchecked((int)0xFF7F7F7F);
        public static int Black = unchecked((int)0xFF000000);

        public static int random(int color, int delta)
        {
            Random rnd = new Random();

            int r = (color >> 16) & 0xff;
            int g = (color >> 8) & 0xff;
            int b = color & 0xff;

            r += (int)(rnd.NextDouble() * (float)delta);
            g += (int)(rnd.NextDouble() * (float)delta);
            b += (int)(rnd.NextDouble() * (float)delta);

            return getCropColor(r, g, b);
        }

        public static int random()
        {
            Random rnd = new Random();
            return (int)(rnd.NextDouble() * 16777216);
        }

        [System.Runtime.CompilerServices.MethodImpl(System.Runtime.CompilerServices.MethodImplOptions.AggressiveInlining)]
        public static int getAlpha(int c)
        {
            return (c >> 24) & 0xff;
        }

        [System.Runtime.CompilerServices.MethodImpl(System.Runtime.CompilerServices.MethodImplOptions.AggressiveInlining)]
        public static int getRed(int c)
        {
            return (c >> 16) & 0xff;
        }

        [System.Runtime.CompilerServices.MethodImpl(System.Runtime.CompilerServices.MethodImplOptions.AggressiveInlining)]
        public static int getGreen(int c)
        {
            return (c >> 8) & 0xff;
        }

        [System.Runtime.CompilerServices.MethodImpl(System.Runtime.CompilerServices.MethodImplOptions.AggressiveInlining)]
        public static int getBlue(int c)
        {
            return c & 0xff;
        }

        public static int[] makeGradient(int[] colors, int size)
        {
            int[] pal = new int[size];
            int c1, c2, pos1, pos2, range;
            int r, g, b, r1, g1, b1, r2, g2, b2, dr, dg, db;
            if(colors.GetLength(0) == 1)
            {
                c1 = colors[0];
                for(int i = 0; i < size; i++)
                {
                    pal[i] = c1;
                }
                return pal;
            }

            for(int c = 0; c < colors.GetLength(0) - 1; c++)
            {
                c1 = colors[c];
                c2 = colors[c + 1];
                pos1 = size * c / (colors.GetLength(0) - 1);
                pos2 = size * (c + 1) / (colors.GetLength(0) - 1);
                range = pos2 - pos1;
                r1 = warp_Color.getRed(c1) << 16;
                g1 = warp_Color.getGreen(c1) << 16;
                b1 = warp_Color.getBlue(c1) << 16;
                r2 = warp_Color.getRed(c2) << 16;
                g2 = warp_Color.getGreen(c2) << 16;
                b2 = warp_Color.getBlue(c2) << 16;
                dr = (r2 - r1) / range;
                dg = (g2 - g1) / range;
                db = (b2 - b1) / range;
                r = r1;
                g = g1;
                b = b1;
                for(int i = pos1; i < pos2; i++)
                {
                    pal[i] = getColor(r >> 16, g >> 16, b >> 16);
                    r += dr;
                    g += dg;
                    b += db;
                }
            }

            return pal;
        }

        [System.Runtime.CompilerServices.MethodImpl(System.Runtime.CompilerServices.MethodImplOptions.AggressiveInlining)]
        public static int add(int color1, int color2)
        {
            int t1,t2,t3,t4;

            t1 = color1 & MASKRB;
            t2 = color2 & MASKRB;
            t3 = t1 + t2;
            t3 |= MASKRBOVER - ((t3 >> 8) & MASKRB);
            t3 &= MASKRB;

            t1 = (color1 >> 8) & MASKRB;
            t2 = (color2 >> 8) & MASKRB;
            t4 = t1 + t2;
            t4 |= MASKRBOVER - ((t4 >> 8) & MASKRB);
            t4 &= MASKRB;

            return (t4 << 8) | t3;
        }

        [System.Runtime.CompilerServices.MethodImpl(System.Runtime.CompilerServices.MethodImplOptions.AggressiveInlining)]
        public static int avg2c(int color1, int color2)
        {
            int t1,t2,t3,t4;

            t1 = color1 & MASKRB;
            t2 = color2 & MASKRB;
            t3 = t1 + t2 + 0x00010001;
            t3 >>= 1;
            t3 &= MASKRB;

            t1 = (color1 >> 8) & MASKRB;
            t2 = (color2 >> 8) & MASKRB;
            t4 = t1 + t2 + 0x00010001;
            t4 >>= 1;
            t4 &= MASKRB;

            return (t4 << 8) | t3;
        }

        [System.Runtime.CompilerServices.MethodImpl(System.Runtime.CompilerServices.MethodImplOptions.AggressiveInlining)]
        public static int avg4c(int color1, int color2, int color3, int color4)
        {
            int t1,t2,t3,t4;

            t1 = color1 & MASKRB;
            t2 = color2 & MASKRB;
            t3 = t1 + t2;
            t1 = color3 & MASKRB;
            t2 = color4 & MASKRB;
            t3 += t1 + t2 + 0x00020002;
            t3 >>= 2;
            t3 &= MASKRB;

            t1 = (color1 >> 8) & MASKRB;
            t2 = (color2 >> 8) & MASKRB;
            t4 = t1 + t2 + 0x00020002;;
            t1 = (color3 >> 8) & MASKRB;
            t2 = (color4 >> 8) & MASKRB;
            t4 += t1 + t2;
            t4 >>= 2;
            t4 &= MASKRB;

            return (t4 << 8) | t3;
        }

        [System.Runtime.CompilerServices.MethodImpl(System.Runtime.CompilerServices.MethodImplOptions.AggressiveInlining)]
        public static int sub(int color1, int color2)
        // Substracts color2 from color1
        {

            int t1,t2,t3,t4;

            t1 = color1 & MASKRB;
            t2 = (~color2) & MASKRB;
            t3 = t1 + t2;
            t3 |= MASKRBOVER - ((t3 >> 8) & MASKRB);
            t3 &= MASKRB;

            t1 = (color1 >> 8) & MASKRB;
            t2 = ((~color2) >> 8) & MASKRB;
            t4 = t1 + t2;
            t4 |= MASKRBOVER - ((t4 >> 8) & MASKRB);
            t4 &= MASKRB;

            return (t4 << 8) | t3;

/*
            int pixel = (color1 & MASK7Bit) + (~color2 & MASK7Bit);
            int overflow = ~pixel & 0x1010100;
            overflow = overflow - (overflow >> 8);
            return ALPHA | (~overflow & pixel);
*/
        }
/*
        public static int subneg(int color1, int color2)
        // Substracts the negative of color2 from color1
        {
            int pixel = (color1 & MASK7Bit) + (color2 & MASK7Bit);
            int overflow = ~pixel & 0x1010100;
            overflow = overflow - (overflow >> 8);
            return ALPHA | (~overflow & pixel);
        }

        public static int inv(int color)
        // returns the inverse of the given color
        {
            color = (~color) &  RGB;
            return ALPHA | color;
        }
*/
        [System.Runtime.CompilerServices.MethodImpl(System.Runtime.CompilerServices.MethodImplOptions.AggressiveInlining)]
        public static int mix(int color1, int color2)
        // Returns the averidge color from 2 colors
        {
            return MASKALPHA | (((color1 & MASK7Bit) >> 1) + ((color2 & MASK7Bit) >> 1));
        }

        [System.Runtime.CompilerServices.MethodImpl(System.Runtime.CompilerServices.MethodImplOptions.AggressiveInlining)]
        public static int scale(int color, int factor)
        {
            if(factor == 255)
                return color;

            if(factor == 0)
                return 0;

            int a = color & MASKALPHA;

            if(factor == 127)
                return a | ((color >> 1) & 0x7F7F7F);

            int RB = (((color & MASKRB) * factor) >> 8) & MASKRB;
            int b = (((color & MASKBLUE) * factor) >> 8) & MASKBLUE;
            return a | RB | b;
        }

        [System.Runtime.CompilerServices.MethodImpl(System.Runtime.CompilerServices.MethodImplOptions.AggressiveInlining)]
        public static int multiply(int color1, int color2)
        {
            int t1, t2, t3, t4;

            t1 = color1;
            t2 = color2;

            t3 = (t1 & 0xff) * (t2 & 0xff); // b on g
            t3 |= (t1 & MASKRED) * ((t2 >> 16) & 0xff); // r on a
            t3 += RBHALF; // add halfbit
            t3 = (t3 + ((t3 >> 8) & MASKRB)) >> 8; // in rb place
            t3 &= MASKRB;

            t1 = (color1 >> 8) & MASKRB;
            t2 = (color2 >> 8) & MASKRB;

            t4 = (t1 & 0xff) * (t2 & 0xff); // g on g
            t4 |= (t1 & MASKRED) * ((t2 >> 16) & 0xff); // a on a
            t4 += RBHALF; // add halfbit
            t4 = (t4 + ((t4 >> 8) & MASKRB));
            t4 &= MASKAG;

            return t3 | t4;
/*
            int a;
            int r = (color1 >> 24 ) & 0xff;
            int g = (color2 >> 24 ) & 0xff;
            if( r == 0xff && g == 0xff)
                a = ALPHA;
            else
            {
                a = ((r * g)) & 0xff00;
                a <<= 16;
            }

            if((color1 & RGB) == 0 && (color2 & RGB) == 0)
                return a;

            r = unchecked((((color1 >> 16) & 0xff) * ((color2 >> 16) & 0xff)) & 0xff00);
            g = (((color1 >> 8) & 0xff) * ((color2 >> 8) & 0xff)) & 0xff00;
            int b = ((color1 & 0xff) * (color2 & 0xff)) >> 8;
            return a | r << 8 | g | b;
*/
        }

        [System.Runtime.CompilerServices.MethodImpl(System.Runtime.CompilerServices.MethodImplOptions.AggressiveInlining)]
        public static int transparency(int bkgrd, int color, int alpha)
        {
            if(alpha == 0)
            {
                return bkgrd;
            }
            if(alpha == 255)
            {
                return color;
            }
            if(alpha == 127)
            {
                return mix(bkgrd, color);
            }

            int colorRB = bkgrd & 0xff00ff;
            colorRB += (((color & 0xff00ff) - colorRB) * alpha) >> 8;
            colorRB &= 0xff00ff;

            int colorg = bkgrd & 0x00ff00;
            colorg += (((color & 0x00ff00) - colorg) * alpha) >> 8;
            colorg &= 0x00ff00;

            return MASKALPHA | colorRB | colorg;
        }

        [System.Runtime.CompilerServices.MethodImpl(System.Runtime.CompilerServices.MethodImplOptions.AggressiveInlining)]
        // blends color alpha, but result image is solid
        public static int overSolid(int bkgrd, int color)
        {
            int alpha = color & MASKALPHA;
            if(alpha == 0)
            {
                return bkgrd;
            }
            if(alpha == Black)
            {
                return color;
            }
            if(alpha == MASKHALFALPHA)
            {
                return mix(bkgrd, color);
            }
            alpha = (alpha >> 24) & 0xff;
            int colorRB = bkgrd & MASKRB;
            colorRB += (((color & MASKRB) - colorRB) * alpha) >> 8;
            colorRB &= 0xff00ff;

            int colorg = bkgrd & MASKGREEN;
            colorg += (((color & MASKGREEN) - colorg) * alpha) >> 8;
            colorg &= MASKGREEN;

            return MASKALPHA | colorRB | colorg;
        }

        [System.Runtime.CompilerServices.MethodImpl(System.Runtime.CompilerServices.MethodImplOptions.AggressiveInlining)]
        public static int getCropColor(int r, int g, int b)
        {
            return MASKALPHA | (warp_Math.crop(r, 0, 255) << 16) | (warp_Math.crop(g, 0, 255) << 8) | warp_Math.crop(b, 0, 255);
        }

        [System.Runtime.CompilerServices.MethodImpl(System.Runtime.CompilerServices.MethodImplOptions.AggressiveInlining)]
        public static int getCropColor(int r, int g, int b, int a)
        {
            return (warp_Math.crop(a , 0, 255) << 24) |
                    (warp_Math.crop(r, 0, 255) << 16) |
                    (warp_Math.crop(g, 0, 255) << 8) |
                    warp_Math.crop(b, 0, 255);
        }

        [System.Runtime.CompilerServices.MethodImpl(System.Runtime.CompilerServices.MethodImplOptions.AggressiveInlining)]
        public static int getColor(int r, int g, int b)
        {
            return MASKALPHA | (r << 16) | (g << 8) | b;
        }

        [System.Runtime.CompilerServices.MethodImpl(System.Runtime.CompilerServices.MethodImplOptions.AggressiveInlining)]
        public static int getColor(int r, int g, int b, int a)
        {
            return (a << 24) | (r << 16) | (g << 8) | b;
        }

        [System.Runtime.CompilerServices.MethodImpl(System.Runtime.CompilerServices.MethodImplOptions.AggressiveInlining)]
        public static int getGray(int color)
        {
            int r = ((color & MASKRED) >> 16);
            int g = ((color & MASKGREEN) >> 8);
            int b = (color & MASKBLUE);
            int Y = (r * 3 + g * 6 + b) / 10;
            return (color & MASKALPHA) | (Y << 16) | (Y << 8) | Y;
        }

        [System.Runtime.CompilerServices.MethodImpl(System.Runtime.CompilerServices.MethodImplOptions.AggressiveInlining)]
        public static int getAverage(int color)
        {
            return (((color & MASKRED) >> 16) + ((color & MASKGREEN) >> 8) + (color & MASKBLUE)) / 3;
        }
    }
}
